src/Symfony/Validator/Validator.php

```diff
@@ -25,19 +26,34 @@
  */
 final class Validator implements ValidatorInterface
 {
-    use ValidationGroupsExtractorTrait;
-
-    public function __construct(private readonly SymfonyValidatorInterface $validator, ?ContainerInterface $container = null)
+    public function __construct(private readonly SymfonyValidatorInterface $validator, private readonly ?ContainerInterface $container = null)
     {
-        $this->container = $container;
     }
 
     /**
      * {@inheritdoc}
      */
     public function validate(object $data, array $context = []): void
     {
-        $violations = $this->validator->validate($data, null, $this->getValidationGroups($context['groups'] ?? null, $data));
+        if (null !== $validationGroups = $context['groups'] ?? null) {
+            if (
+                $this->container
+                && \is_string($validationGroups)
+                && $this->container->has($validationGroups)
+                && ($service = $this->container->get($validationGroups))
+                && \is_callable($service)
+            ) {
+                $validationGroups = $service($data);
+            } elseif (\is_callable($validationGroups)) {
+                $validationGroups = $validationGroups($data);
+            }
+
+            if (!$validationGroups instanceof GroupSequence) {
+                $validationGroups = (array) $validationGroups;
```

Comment:
But passing array of groups is not the same as passing GroupSequence 🤔

Reply:
This is just a revert of changes made in https://github.com/api-platform/core/pull/7184

This condition says - if `$validationGroups` is not an instance of GroupSequence, make it an array.

Reply:
Oh right,read it wrong 🤦‍♂️

Reply:
Yes, yes ) Yesterday I did the same ))